wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/convert/tool/kubernetes/PodsMerger.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.devfile.server.convert.tool.kubernetes;
+
+import io.fabric8.kubernetes.api.model.ObjectMeta;
+import io.fabric8.kubernetes.api.model.ObjectMetaBuilder;
+import io.fabric8.kubernetes.api.model.PodSpec;
+import io.fabric8.kubernetes.api.model.PodSpecBuilder;
+import java.util.List;
+import java.util.Optional;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment.PodData;
+
+/**
+ * Helps to merge multiple pods to single deployment.
+ *
+ * @author Sergii Leshchenko
+ */
+public class PodsMerger {
+
+  /**
+   * Merges the specified pods list into one single deployment.
+   *
+   * @param pods list to merge
+   * @return created deployment with pod template which is a result of a merge
+   * @throws IllegalArgumentException if the specified pods list does not contain any pod data
+   */
+  public static PodData merge(List<PodData> pods) {
+    Optional<PodData> merged = pods.stream().reduce(PodsMerger::merge);
+
+    if (!merged.isPresent()) {
+      // merged pod is absent it means that source pods list is empty
+      throw new IllegalArgumentException("At least one pods data is expected");
+    }
+
+    return merged.get();
+  }
+
+  private static PodData merge(PodData podData1, PodData podData2) {
+    ObjectMeta mergedMeta = merge(podData1.getMetadata(), podData2.getMetadata());
+    PodSpec mergedSpec = merge(podData1.getSpec(), podData2.getSpec());
+    return new PodData(mergedSpec, mergedMeta);
+  }
+
+  private static ObjectMeta merge(ObjectMeta meta1, ObjectMeta meta2) {
+    ObjectMeta result = new ObjectMetaBuilder(meta1).build();
+    // T_O_D_O What to do if both metas contain labels with the same keys?
+    // We can not just override it and use any value because it influences
+    // picking up pods by Services
+    // Since Workspace can contain only one Pod - we should match all services for that pod
+    // like by setting one const label
+    result.getLabels().putAll(meta2.getLabels());
+    result.getAnnotations().putAll(meta2.getAnnotations());
+    result.getAdditionalProperties().putAll(meta2.getAdditionalProperties());
+    return result;
+  }
+
+  private static PodSpec merge(PodSpec podSpec1, PodSpec podSpec2) {
+    PodSpec result = new PodSpecBuilder(podSpec1).build();
+    result.getContainers().addAll(podSpec2.getContainers());
```

Comment:
How do we handle name collisions here?

Reply:
We do not handle name collisions here.

Reply:
Do we handle them somewhere? :)

Reply:
No) Now if a recipe contains objects with the same names (Pod, Service) we will use only last =(
If pod contains containers with a not unique name - workspace start will fail because k8s/os API will response an error.

In this PR I can handle containers names collisions because a user doesn't carry about real container name (at least in most cases, if he does not rely on pod DNS). So, I can add some suffix in container name if needed.
As about pod/services/configmaps/secrets/(volumes inside pods) names - I will throw an exception and create a separate issue to resolving conflicts here since it requires more work, like cascade changing all objects.